vernac/comDefinition.ml

```diff
@@ -48,57 +48,59 @@ let warn_implicits_in_term =
           strbrk "Implicit arguments declaration relies on type." ++ spc () ++
             strbrk "The term declares more implicits than the type here.")
 
+let check_imps ~impsty ~impsbody =
+  let b =
+    try
+      List.for_all (fun (key, (va:bool*bool*bool)) ->
+          (* Pervasives.(=) is OK for this type *)
+          Pervasives.(=) (List.assoc_f Impargs.explicitation_eq key impsty) va)
+        impsbody
+    with Not_found -> false
+  in
+  if not b then warn_implicits_in_term ()
+
 let interp_definition pl bl poly red_option c ctypopt =
+  let open EConstr in
   let env = Global.env() in
+  (* Explicitly bound universes and constraints *)
   let evd, decl = Univdecls.interp_univ_decl_opt env pl in
+  (* Build the parameters *)
   let evd, (impls, ((env_bl, ctx), imps1)) = interp_context_evars env evd bl in
-  let nb_args = Context.Rel.nhyps ctx in
-  let evd,imps,ce =
-    match ctypopt with
-      None ->
-        (* Closing a definition, first build the term, this is straightforward *)
-        let env_bl = EConstr.push_rel_context ctx env in
-        let evd, (c, imps2) = interp_constr_evars_impls ~impls env_bl evd c in
-        let body = EConstr.(it_mkLambda_or_LetIn c ctx) in
-        (* Second, we need to prepare the universes, including minimization. Not fun *)
-        let evd,_ = Evarutil.nf_evars_and_universes evd in
-        let vars = EConstr.universes_of_constr env evd body in
-        let evd = Evd.restrict_universe_context evd vars in
-        let uctx = Evd.check_univ_decl ~poly evd decl in
-        (* Third, we substitute all the evars to store in the kernel. *)
-        let body = EConstr.to_constr evd body in
-        (* Fourth, store *)
-        evd, imps1@(Impargs.lift_implicits nb_args imps2),
-          definition_entry ~univs:uctx body
-    | Some ctyp ->
-        (* 1 *)
-        let evd, (ty, impsty) = interp_type_evars_impls ~impls env_bl evd ctyp in
-        let env_bl = EConstr.push_rel_context ctx env in
-        let evd, (c, imps2) = interp_casted_constr_evars_impls ~impls env_bl evd c ty in
-        let body = EConstr.(it_mkLambda_or_LetIn c ctx) in
-        let typ = EConstr.(it_mkProd_or_LetIn ty ctx) in
```

Comment:
Here we didn't shadow `ty` which I guess let us avoid folding over the parameters twice in universes_of_constr. I don't believe this will show up in a benchmark so I removed this "optimization" (might have been just a typo).

Reply:
>Here we didn't shadow ty which I guess let us avoid folding over the parameters twice in universes_of_constr. I don't believe this will show up in a benchmark so I removed this "optimization" (might have been just a typo).

Maybe this did show up then. I'll test.